aiida/transports/plugins/ssh.py

```diff
@@ -1255,14 +1258,68 @@ def exec_command_wait(self, command, stdin=None, combine_stderr=False, bufsize=-
         ssh_stdin.flush()
         ssh_stdin.channel.shutdown_write()
 
+        # Now I get the output
+        stdout_bytes = []
+        stderr_bytes = []
+        # 10kB buffer (note that this should be smaller than the window size of paramiko)
+        # Also, apparently if the data is coming slowly, the read() command will not unlock even for
+        # times much larger than the timeout. Therefore we don't want to have very large buffers otherwise
+        # you risk that a lot of output is sent to both stdout and stderr, and stderr goes beyond the
+        # winodw size and blocks.
+        bufsize = 10 * 1024
+
+        # Set a small timeout on the channels, so that if we get data from both
+        # stderr and stdout, and the connection is slow, we interleave the receive and don't hang
+        stdout.channel.settimeout(0.1)
+        stderr.channel.settimeout(0.1)  # Maybe redundant, as this could be the same channel.
+
+        while True:
+            chunk_exists = False
+
+            if stdout.channel.recv_ready():  # True means that the next .read call will at least receive 1 byte
+                chunk_exists = True
+                try:
+                    piece = stdout.read(bufsize)
+                    stdout_bytes.append(piece)
+                except socket.timeout:
```

Comment:
Is there any possibility that `socket.timeout` will be raised while `std*.channel.recv_stderr_ready()` keeps returning `True`? That could be an infinite loop.

Reply:
From the [current docs](http://docs.paramiko.org/en/stable/api/channel.html):
- The documentation of `recv_ready` says that it returns True if a `recv` call on this channel would immediately return at least one byte; False otherwise.
- The documentation of `recv` says that it raises socket.timeout if no data is ready before the timeout set by settimeout.

What I gather is one of the two options:
- if a timeout is reached when only `nbytes < bufsize` bytes are received (and the channel is still open), just return those `nbytes>0`
- if a timeout is reached when only `nbytes < bufsize` bytes are received (and the channel is still open), raise. The bytes will be available in the next `recv` call if more arrive in the buffer and they go > `bufsize`, or if the channel is closed.

In both cases, this should not give an infinite loop.

Actually, from what I remember from my tests while implementing, there could even be a different possibility: I remember there was one case where no timeout was being raised, but still the data was arriving *very* slowly (with the call blocking on `recv` for much longer than the timeout, i.e., seconds, even if the bufsize was small). 
I am not sure how to interpret this. Maybe there is some kind of heartbeat in the connection, or maybe bytes were sent one at a time very slowly? However, I am not even sure I would be able to reproduce this now...